Load Output Properties Fix

Matt-Cumber opened the pull request:
**Description of work.**
I have updated Load to now output properties of all types not just workspaces. This allows us to use load in the Muon guis now.

**To test:**
Run this script to test load outputting properties:
```
# Load Single file
load_alg = AlgorithmManager.create("Load")
load_alg.initialize()
load_alg.setChild(True)
load_alg.setProperty("Filename", "MUSR62260")
load_alg.setProperty("OutputWorkspace", "OutWs")
load_alg.setProperty("SpectrumMin", 20)
load_alg.setProperty("SpectrumMax", 32)
load_alg.setProperty("DeadTimeTable", "dead_time_table")
load_alg.setProperty("DetectorGroupingTable", "detector_grouping_table")
load_alg.setProperty("TimeZeroTable", "time_zero_table")
load_alg.execute()

print("From Load")
time_zero = load_alg.getProperty("TimeZero")
print(time_zero.value)
field = load_alg.getProperty("MainFieldDirection")
print(field.value)
first_good_data = load_alg.getProperty("FirstGoodData")
print(first_good_data.value)
dead_time_table = load_alg.getProperty("DeadTimeTable")
if dead_time_table.value:
    for row in dead_time_table.value:
        print(row)
detector_grouping_table = load_alg.getProperty("DetectorGroupingTable")
if detector_grouping_table.value:
    for row in detector_grouping_table.value:
        print(row)
                
# Compare to using other loaders should be same result
print("")

load_alg = AlgorithmManager.create("LoadMuonNexus")
load_alg.initialize()
load_alg.setChild(True)
load_alg.setProperty("Filename", "MUSR62260")
load_alg.setProperty("OutputWorkspace", "OutWsLoadMuonNexus")
load_alg.setProperty("SpectrumMin", 20)
load_alg.setProperty("SpectrumMax", 32)
load_alg.setProperty("DeadTimeTable", "dead_time_table")
load_alg.setProperty("DetectorGroupingTable", "detector_grouping_table")
load_alg.setProperty("TimeZeroTable", "time_zero_table")
load_alg.execute()
 
print("From LoadMuonNexus")
time_zero = load_alg.getProperty("TimeZero")
print(time_zero.value)
field = load_alg.getProperty("MainFieldDirection")
print(field.value)
first_good_data = load_alg.getProperty("FirstGoodData")
print(first_good_data.value)
dead_time_table = load_alg.getProperty("DeadTimeTable")
if dead_time_table.value:
    for row in dead_time_table.value:
        print(row)
detector_grouping_table = load_alg.getProperty("DetectorGroupingTable")
if detector_grouping_table.value:
    for row in detector_grouping_table.value:
        print(row)
```

1. Should also check multi-period data and multi-files for the load alg (A multi-period file is HIFI110542)
2. Open Muon Analysis
3. You will need access to a PSI data file, and a nexus file in nexus v1 and v2 formats (Use a new nexusv2 file as old files do not work since they are in the wrong format, I have been using EMU00102347)
4. To load PSI data, change instrument to PSI and browse to a PSI file (e.g. deltat_tdc_dolly_1529.bin)
5. This should loas successfully
6. Now load the nexusv1 file, this should work
7. Clear the current data and browse to the nexusv2 file (should be exactly the same as the v1 file)

Fixes #29715 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Matt-Cumber committed:
Re 29715 Basic LoadMuonData Alg

Matt-Cumber committed:
Re #29715 LoadMuonData Now Cleaner

Matt-Cumber committed:
Re #29715 Added Test File

AnthonyLim23 commented on Framework/DataHandling/test/LoadMuonDataTest.h:
```diff
@@ -0,0 +1,34 @@
+  void testExecWithBinFile() {
+    loader.initialize();
+    loader.setProperty("Filename", "deltat_tdc_dolly_1529.bin");
+    loader.setProperty("OutputWorkspace", "OutWS");
```
can we check the outputs too (at least they are what we expect, such a a deadtime table)?

Matt-Cumber committed:
Re #29715 Complete LoadMuonData Alg

AnthonyLim23 commented:
I just hacked it into the GUI.

PSI worked fine

nxs gave an error ```Error in execution of algorithm LoadMuonNexus
Add Data Object with empty name````

AnthonyLim23 commented:
change https://github.com/mantidproject/mantid/blob/master/scripts/Muon/GUI/Common/load_file_widget/presenter.py#L50 to ```         file_filter = file_utils.filter_for_extensions(file_utils.allowed_extensions)```

and change https://github.com/mantidproject/mantid/blob/master/scripts/Muon/GUI/Common/utilities/muon_file_utils.py#L12
to ``` allowed_extensions = ["nxs", "bin", "nxs_v2"]```

and change https://github.com/mantidproject/mantid/blob/master/scripts/Muon/GUI/Common/utilities/load_utils.py#L242 to 
```
def create_load_algorithm(filename, property_dictionary):
    # Assume if .bin it is a PSI file
    psi_data = False
    output_filename = os.path.basename(filename)
    alg = mantid.AlgorithmManager.create("LoadMuonData")
    alg.setProperties(property_dictionary)

    if ".bin" in filename:
        psi_data = True
    else:
        alg = mantid.AlgorithmManager.create("LoadMuonNexus")
    alg.setProperty("DeadTimeTable", output_filename + '_deadtime_table')
```

Matt-Cumber committed:
Re #29715 Added Alg To MA And FDA

Matt-Cumber committed:
Re #29715 Testing Using Load Alg Instead

Matt-Cumber committed:
Re #29715 Remove LoadMuonData Alg

Matt-Cumber committed:
Re #29715 Load Alg Is Now Used To Load Data In Muon GUIs

Matt-Cumber committed:
Re #29715 Added Check For Value On Properties Before Setting

This fixes added empty data object error. Also re-include output workspace property setter to handle multi-period properly

Matt-Cumber committed:
Re #29715 New Method takeProperty

Use this new method to transfer ownership of property pointers to parent load alg from child loader

AnthonyLim23 commented:
can we add a system test that checks the signatures match for the 3 muon loaders?

AnthonyLim23 commented:
the script works for all 3 muon loaders (with some minor changes) and I can load all 3 data types into the muon GUIs.

Would it be much effort to update the ALC too or should that be a different job?

Matt-Cumber commented:
> Would it be much effort to update the ALC too or should that be a different job?

The ALC uses LoadMuonNexus to populate the interface, and then once Load is pressed uses the PlotAsymmetryByLogValue algorithm to loaded the data which also uses LoadMuonNexus as a child alg. I feel like this should be a separate PR

AnthonyLim23 commented:
> > Would it be much effort to update the ALC too or should that be a different job?
> 
> The ALC uses LoadMuonNexus to populate the interface, and then once Load is pressed uses the PlotAsymmetryByLogValue algorithm to load the data which also uses LoadMuonNexus as a child alg. I feel like this should be a separate PR

ok, could you make an issue for it?

Matt-Cumber committed:
Re #29715 Added LastGoodData

Added last good data to muon loaders so signatures match. Added system test to test signatures of loaders

Matt-Cumber committed:
Re #29715 Handle OutputWorkspace Type Problems

Matt-Cumber committed:
Re #29715 Fix Unit Tests

AnthonyLim23 commented:
the nxs 2 files dont work with the period info. I used HIFI 103767

AnthonyLim23 commented on Framework/DataHandling/src/LoadMuonNexus1.cpp:
```diff
@@ -111,6 +111,33 @@ void LoadMuonNexus1::exec() {
+  try {
+    NXInfo infoResolution = entry.getDataSetInfo("resolution");
+    NXInt counts = root.openNXInt("run/histogram_data_1/counts");
+    std::string lastGoodBin = counts.attributes("last_good_bin");
```
there is no last_good_bin entry in the nxs 1 files. So just set the last good value to be the largest x value. And add a comment in the code explaining that there isnt a value in the files.

AnthonyLim23 commented on Framework/DataHandling/src/LoadMuonNexus2.cpp:
```diff
@@ -234,6 +234,11 @@ void LoadMuonNexus2::doExec() {
     }
+
+    if (detector.containsDataSet("last_good_time")) {
+      double dum = detector.getFloat("last_good_time");
```
this is not in the nxs2 files either. So lets just set it to the largest x value

AnthonyLim23 commented on Framework/DataHandling/src/Load.cpp:
```diff
@@ -521,12 +521,88 @@ void Load::setUpLoader(const API::IAlgorithm_sptr &loader, const double startPro
+}
+
+API::Workspace_sptr Load::getOutputWorkspace(const std::string &propName, const API::IAlgorithm_sptr &loader) const {
+  // @todo Need to try and find a better way using the getValue methods
```
do you want to leave this here?

Matt-Cumber replied:
I think we do want to leave this here. It would be nice to have a better way of setting the output workspace but it is difficult due to the various types of workspaces we can have. Since this PR already got taken out of scope, I think it is best to leave this in for the future and stick to just fixing the initial problem

Matt-Cumber committed:
Re #29715 Change How Output Workspace Set

There were problems arising from the output workspace. To avoid this, this property is set on its own with the remaining properties set in the setOutputProperties method. Also added an extra release note for the muon alg changes

AnthonyLim23 commented:
Open muon analysis
Load HIFI 84447
Turn off dead times
Then use browse to open a HIFI nxs2 file
Crash

AnthonyLim23 commented:
> Open muon analysis
> Load HIFI 84447
> Turn off dead times
> Then use browse to open a HIFI nxs2 file
> Crash

this is fixed in https://github.com/mantidproject/mantid/pull/32062

AnthonyLim23 reviewed (dismissed):
It allows us to use Load for any muon data, making the GUI much cleaner.

AnthonyLim23 reviewed (dismissed):
All of the muon stuff works as expected

martyngigg committed:
Add check for output properties set by Load

This also removes a duplicated test that was outdated
after the AlgorithmProxy class was removed
